XMonad/Actions/EasyMotion.hs

```diff
@@ -0,0 +1,332 @@
+           }
+
+-- | Use to create overlay windows the same size as the window they select
+fullSize :: Position -> Rectangle -> Rectangle
+fullSize _ = id
+
+-- | Use to create overlay windows a proportion of the size of the window they select
+proportional :: RealFrac f => f -> Position -> Rectangle -> Rectangle
+proportional f th r = Rectangle { rect_width  = newW
+                                , rect_height = newH
+                                , rect_x      = rect_x r + fi (rect_width r - newW) `div` 2
+                                , rect_y      = rect_y r + fi (rect_height r - newH) `div` 2 }
+                                  where
+                                    newH = max (fi th) (round $ f * fi (rect_height r))
+                                    newW = newH
+
+-- | Use to create fixed-size overlay windows
+fixedSize :: (Integral a, Integral b) => a -> b -> Position -> Rectangle -> Rectangle
+fixedSize w h th r = Rectangle { rect_width  = rw
+                               , rect_height = rh
+                               , rect_x      = rect_x r + fi (rect_width r - rw) `div` 2
+                               , rect_y      = rect_y r + fi (rect_height r - rh) `div` 2 }
+                                 where
+                                   rw = max (fi w) (fi th)
+                                   rh = max (fi h) (fi th)
+
+-- | Use to create overlay windows the minimum size to contain their key chord
+textSize :: Position -> Rectangle -> Rectangle
+textSize th r = Rectangle { rect_width  = fi th
+                          , rect_height = fi th
+                          , rect_x      = rect_x r + (fi (rect_width r) - fi th) `div` 2
+                          , rect_y      = rect_y r + (fi (rect_height r) - fi th) `div` 2 }
+
+-- | Use to create overlay windows the full width of the window they select, the minimum height to
+--   contain their chord, and a proportion of the distance from the top of the window they select
+bar :: RealFrac f => f -> Position -> Rectangle -> Rectangle
+bar f th r = Rectangle { rect_width  = rect_width r
+                       , rect_height = fi th
+                       , rect_x      = rect_x r
+                       , rect_y      = rect_y r + round (f' * (fi (rect_height r) - fi th)) }
+                         -- clamp f in [0,1] as other values will appear to lock up xmonad
+                         -- as the overlay will be displayed off-screen
+                         where f' = min 0.0 $ max f 1.0
+
+-- | Handles overlay display and window selection. Called after config has been sanitised.
+handleSelectWindow :: EasyMotionConfig -> X (Maybe Window)
+handleSelectWindow c = do
+  f <- initXMF $ font c
+  th <- textExtentsXMF f (concatMap keysymToString (concat $ sKeys c)) >>= \(asc, dsc) -> return $ asc + dsc + 2
+  XConf { theRoot = rw, display = dpy } <- ask
+  XState { mapped = mappedWins, windowset = ws } <- get
+  let currentW = W.stack . W.workspace . W.current $ ws
```

Comment:
Because you're only using `currentW` to get to the current focus below, you can do 
```suggestion
  let currentW = W.peek $ ws
```